src/core/stdcpp/array.d

```diff
@@ -97,7 +97,7 @@ pure nothrow @nogc:
         ///
         inout(T)* data() inout @safe                    { static if (N > 0) { return &_M_elems[0]; } else { return null; } }
         ///
-        ref inout(T[N]) as_array() inout @trusted       { return data()[0 .. N]; }
+        ref inout(T)[N] as_array() inout @trusted       { return cast(inout(T)[N])data()[0 .. N]; }
```

Comment:
I feel like that cast should never be necessary. Could you explain why it is ?

Reply:
data returns an inout(T)* which cannot be implicitly converted to ref inout(T)[N] because you do not have a size.  Note that the function is @trusted because the compiler introduced the cast implicitly before the patch in https://github.com/dlang/dmd/pull/9505 .

Reply:
Why don't you access `_M_elems` directly, which already has the correct type?

Reply:
I wanted the change to be minimally invasive.

Reply:
I think it's better for it to be correct and avoiding casts.